numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -47,11 +47,12 @@ PyArray_CastToType(PyArrayObject *arr, PyArray_Descr *dtype, int is_f_order)
     PyObject *out;
 
     /* If the requested dtype is flexible, adapt it */
-    PyArray_AdaptFlexibleDType((PyObject *)arr, PyArray_DESCR(arr), &dtype);
-    if (dtype == NULL) {
-        return NULL;
+    if (dtype != NULL) {
+        dtype = PyArray_AdaptFlexibleDType((PyObject *)arr, PyArray_DESCR(arr), dtype);
+        if (dtype == NULL) {
+            return NULL;
+        }
```

Comment:
Hmm, seems ok, did I ask for this to be changed? Seems that dtype cannot be possibly NULL in this function, or things would go down in flames?

Reply:
removed exuberant check